lib/graphql_connector/service_classable/query.rb

```diff
@@ -0,0 +1,104 @@
+# frozen_string_literal: true
+
+module GraphqlConnector
+  module ServiceClassable
+    class InvalidTypeMappingError < StandardError
+    end
+    class InvalidParamsErrors < StandardError
+    end
+
+    class ReturnFieldsErrors < StandardError
+    end
+    # Base module that is invoked in BaseServerType
+    module Query
+      def self.included(base)
+        base.extend Queryable
+      end
+
+      # Module that allows to build query methods within the context of
+      # service class
+      module Queryable
+        CONDITIONS = 'binding.local_variables.map do |var|
+                        [var, binding.local_variable_get(var)]
+                      end.to_h'
+
+        def return_fields(*fields)
+          @return_fields ||= Set.new
+          @return_fields.merge(fields)
+        end
+
+        def query(**type_mapping)
+          params = type_mapping.delete(:params) || {}
+          method_name = type_mapping.first[0]
+          query_type = type_mapping.first[1]
+          raw_query = query_type.is_a?(String)
+          validate!(type_mapping, params)
+
+          if params.empty?
+            return query_method(method_name, query_type, raw_query)
+          end
+
+          keyword_query_method(method_name, query_type, params, raw_query)
+        end
+
+        private
+
+        def query_method(name, query_type, raw_query)
+          define_singleton_method name do
+            return http_client.raw_query(query_type) if raw_query
+
+            http_client.query(query_type, {}, @return_fields.to_a)
+          end
+        end
+
+        def keyword_query_method(name, query_type, params, raw_query)
+          params = [params].flatten
+          instance_eval <<-METHOD, __FILE__, __LINE__ + 1
+            def #{name}(#{params.map { |keyword| "#{keyword}:" }.join(', ')})
+              if #{raw_query}
+                http_client
+                  .raw_query("#{query_type}", variables: #{CONDITIONS})
+              else
+                http_client
+                  .query("#{query_type}", #{CONDITIONS}, #{@return_fields.to_a})
+              end
+            end
+          METHOD
+        end
+
+        def validate!(type_mapping, params)
+          validate_mapping!(type_mapping)
+          validate_params!(params) unless params.empty?
+          validate_return_fields!
+        end
+
+        def validate_mapping!(type_mapping)
+          return if type_mapping.size == 1 &&
+                    type_mapping.first[0].is_a?(Symbol) &&
+                    (type_mapping.first[1].is_a?(Symbol) ||
```

Comment:
What do you think about:
```ruby
[String, Symbol].member?(type_mapping.first[1].class)
```

And second thing does it make sense to support string a symbols for the second parameter, but not for the first one?
```ruby
'test' => 'blubber'
```

Reply:
> What do you think about:
> [String, Symbol].member?(type_mapping.first[1].class)

Yes that looks more readable. Changed it to that

> And second thing does it make sense to support string a symbols for the second parameter, but not for the first one?
> 'test' => 'blubber'

That is a good question. However we cannot use this syntax if want to keep the keyword argument with splat parameters in [`def query(**type_mapping`)](https://github.com/Garllon/graphql_connector/pull/14/files#diff-d14d48c082efaf086801f55333ed8758R26)

Reply:
I actually refactored the whole procedure that `String` is also covered (as in your example)

Reply:
Further refactorings:

* Renamed `query` to `add_query` (feels more accurate)
* Added `add_raw_query` which deals with raw graphql queries
* Moved `return_fields` into actual `add_query` segment since such a service class can have multiple `graphql` types and therefore a global `return_fields` would have made no sense